### PR TITLE
sys-fs/ncdu-bin, dev-lang/zig-bin: new packages

### PR DESCRIPTION
**Pull request by BratishkaErik**
Signed-off-by: Eric Joldasov <bratishkaerik@getgoogleoff.me>

### CODE REVIEW

**Comment by BratishkaErik** [comment]
cc @jsmolic: I added you as maintainer because you are maintainer of `sys-fs/ncdu` package, I will drop this change, if you'd like that.

**Comment by jsmolic** [comment]
You should split commits atomically (one commit per package), e.g.in https://github.com/gentoo/gentoo/pull/26325/commits/61880050af762e5fad9b4d9b9ea8ed1568f7bd45 you're also changing dev-lang/zig and adding a virtual package, but this should actually be split in 3 commits, second commits can also be split in 2

**Comment by BratishkaErik** [comment]
> You should split commits atomically (one commit per package), e.g.in https://github.com/gentoo/gentoo/pull/26325/commits/61880050af762e5fad9b4d9b9ea8ed1568f7bd45 you're also changing dev-lang/zig and adding a virtual package, but this should actually be split in 3 commits, second commits can also be split in 2

Sure!

**Comment by BratishkaErik** [comment]
> You should split commits atomically (one commit per package), e.g.in https://github.com/gentoo/gentoo/pull/26325/commits/61880050af762e5fad9b4d9b9ea8ed1568f7bd45 you're also changing dev-lang/zig and adding a virtual package, but this should actually be split in 3 commits, second commits can also be split in 2

Done.

File: dev-lang/zig-bin/zig-bin-0.9.1.ebuild
Comment (jsmolic) [inline dev-lang/zig-bin/zig-bin-0.9.1.ebuild:1]:
```suggestion
# Copyright 2022 Gentoo Authors
```

File: dev-lang/zig-bin/zig-bin-0.9.1.ebuild
Comment (jsmolic) [inline dev-lang/zig-bin/zig-bin-0.9.1.ebuild:13]: Move keywords below SLOT

File: dev-lang/zig/zig-0.9.1.ebuild
Comment (jsmolic) [inline dev-lang/zig/zig-0.9.1.ebuild:26]: I tried to build this on riscv but I still had some issues, although given it's merged upstream we might leave it

File: sys-fs/ncdu-bin/ncdu-bin-2.1.2.ebuild
Comment (jsmolic) [inline sys-fs/ncdu-bin/ncdu-bin-2.1.2.ebuild:1]:
```suggestion
# Copyright 2022 Gentoo Authors
```

File: sys-fs/ncdu-bin/ncdu-bin-2.1.2.ebuild
Comment (jsmolic) [inline sys-fs/ncdu-bin/ncdu-bin-2.1.2.ebuild:8]:
```suggestion
```

File: sys-fs/ncdu-bin/ncdu-bin-2.1.2.ebuild
Comment (jsmolic) [inline sys-fs/ncdu-bin/ncdu-bin-2.1.2.ebuild:21]: Move below SRC_URI

File: sys-fs/ncdu-bin/ncdu-bin-2.1.2.ebuild
Comment (jsmolic) [inline sys-fs/ncdu-bin/ncdu-bin-2.1.2.ebuild:26]: Maybe we can install it as `ncdu-bin`, that way we wouldn't have to have blockers

Comment (BratishkaErik) [reply sys-fs/ncdu-bin/ncdu-bin-2.1.2.ebuild]: Sure!

File: dev-lang/zig/zig-0.9.1.ebuild
Comment (BratishkaErik) [reply dev-lang/zig/zig-0.9.1.ebuild:26]: Is this the same problem as without the patch, or is it different?

Comment (jsmolic) [reply dev-lang/zig/zig-0.9.1.ebuild:26]: I think it was different, but can't remember off the top of my head what it was, so I'll have to try building again with and without the patch.

Comment (BratishkaErik) [reply dev-lang/zig/zig-0.9.1.ebuild:26]: I'm trying to reproduce this with riscv chroot and QEMU user emulation, is it complete enough? (QEMU system emulation is slow for me & I don't have RISC-V hardware)

**Comment by jjdredd** [comment]
If this is merged, It think https://github.com/gentoo/gentoo/pull/26779 can be closed

File: dev-lang/zig/zig-0.9.1.ebuild
Comment (BratishkaErik) [reply dev-lang/zig/zig-0.9.1.ebuild:26]: Well, with this patch `zig` can be builded succesfully, but we can't build ncdu on RISC-V systems yet :(
```sh
$ /tmp/portage/dev-lang/zig-0.9.1/work/zig-0.9.1/build/zig build -Drelease-fast

ld.lld: error: /usr/lib64/lp64d/crt1.o:(.text+0x0): relocation R_RISCV_ALIGN requires unimplemented linker relaxation; recompile with -mno-relax
error: LLDReportedFailure
ncdu...The following command exited with error code 1:
/tmp/portage/dev-lang/zig-0.9.1/work/zig-0.9.1/build/zig build-exe /tmp/portage/sys-fs/ncdu-2.1.2/work/ncdu-2.1.2/src/main.zig /tmp/portage/sys-fs/ncdu-2.1.2/work/ncdu-2.1.2/src/ncurses_refs.c -lc -lncursesw -ltinfow -OReleaseFast --cache-dir /tmp/portage/sys-fs/ncdu-2.1.2/work/ncdu-2.1.2/zig-cache --global-cache-dir /root/.cache/zig --name ncdu -I /usr/include/ncursesw -D _DEFAULT_SOURCE -D _XOPEN_SOURCE=600 --enable-cache
error: the following build command failed with exit code 1:
/tmp/portage/sys-fs/ncdu-2.1.2/work/ncdu-2.1.2/zig-cache/o/23b40c8e0866163e938bfce718fc22b7/build /tmp/portage/dev-lang/zig-0.9.1/work/zig-0.9.1/build/zig /tmp/portage/sys-fs/ncdu-2.1.2/work/ncdu-2.1.2 /tmp/portage/sys-fs/ncdu-2.1.2/work/ncdu-2.1.2/zig-cache /root/.cache/zig -Drelease-fast
```

IMO that's because `zig` misdetects ABI:
```sh
$ /tmp/portage/dev-lang/zig-0.9.1/work/zig-0.9.1/build/zig targets | grep triple

"triple": "riscv64-linux.5.18.15...5.18.15-musl",
```

According to https://maskray.me/blog/2022-07-10-riscv-linker-relaxation-in-lld:
> On 2022-07-07, I added a RISC-V linker relaxation framework in ld.lld and implemented R_RISCV_ALIGN/R_RISCV_CALL/R_RISCV_CALL_PLT relaxation. The changes will be included in the next llvm-project release 15.0.0. This post describes the implementation.

Zig 0.10.0 will be self-hosted by default (see https://github.com/ziglang/zig/pull/12368) and possibly upgraded to LLVM 15, so [I think] we should postpone this until next release...

**Review by jsmolic — COMMENTED** [review]
I can merge the ncdu part but I'd like to have another pair of eyes at zig commits

File: dev-lang/zig-bin/zig-bin-0.9.1.ebuild
Comment (thesamesam) [inline dev-lang/zig-bin/zig-bin-0.9.1.ebuild:18]: I think this should just be ! as it doesn't harm installation at all (and !! isn't needed in RDEPEND anyway, or at least very rarely, could happen with pkg_postinst or something I guess but that should be avoided).

File: dev-lang/zig/files/zig-0.9.1-fix-riscv.patch
Comment (thesamesam) [inline dev-lang/zig/files/zig-0.9.1-fix-riscv.patch:1]: Please add a link to the commit which got merged.

File: dev-lang/zig/zig-0.9.1.ebuild
Comment (thesamesam) [inline dev-lang/zig/zig-0.9.1.ebuild:40]: !

File: dev-lang/zig/zig-9999.ebuild
Comment (thesamesam) [inline dev-lang/zig/zig-9999.ebuild:36]: !

**Review by thesamesam — APPROVED** [review]
LGTM _but_ new virtuals must be sent to the gentoo-dev mailing list.

File: virtual/zig/zig-0.9.1.ebuild
Comment (thesamesam) [inline virtual/zig/zig-0.9.1.ebuild:12]: drop BDEPEND

File: virtual/zig/zig-0.9.1.ebuild
Comment (thesamesam) [inline virtual/zig/zig-0.9.1.ebuild:8]: drop

File: sys-fs/ncdu/ncdu-2.1.2.ebuild
Comment (thesamesam) [inline sys-fs/ncdu/ncdu-2.1.2.ebuild:14]: (This needs a revbump really to allow people to change to -bin, just git mv for it)

File: dev-lang/zig/zig-9999.ebuild
Comment (thesamesam) [inline dev-lang/zig/zig-9999.ebuild:24]: for all commit messages: please no version before the colon (app-misc/foo: blah, but not app-misc/foo-9999: blah. you can do app-misc/foo: blah in 9999 if you want).

**Comment by BratishkaErik** [comment]
> LGTM _but_ new virtuals must be sent to the gentoo-dev mailing list first.

https://archives.gentoo.org/gentoo-dev/message/05a25d66f72bfb3a01ab6c60ed466b5b

**Comment by BratishkaErik** [comment]
Upstream is transitioning to self–hosted compiler, you can ignore build errors for now.

**Review by jsmolic — APPROVED** [review]
Nice work overall, thanks for pushing this to the end, and sorry for the delays in review

**Comment by BratishkaErik** [comment]
> Nice work overall, thanks for pushing this to the end, and sorry for the delays in review

No problems and thank you for helpful reviews!

**Comment by jjdredd** [comment]
Sweet!

**Pull request closed without merging**
